src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs

```diff
@@ -60,28 +82,88 @@ internal ulong BucketSize
                 }
             }
 
-            // Relative cost of the bucket
-            // roughly proportional to the number of compares it needs in the success case.
-            internal int BucketCost
+            // if a bucket could be split into two degenerate ones
+            // specifies a label index where that split would occur
+            // -1 indicates that the bucket cannot be split into degenerate ones
+            //  0 indicates that the bucket is already degenerate
+            //
+            //  Degenerate buckets here are buckets with contiguous range of constants
+            //  leading to the same label. Like:
+            //
+            //      case 0:
+            //      case 1:
+            //      case 2:
+            //      case 3:
+            //           DoOneThing();
+            //           break;               
+            //
+            //      case 4:
+            //      case 5:
+            //      case 6:
+            //      case 7:
+            //           DoAnotherThing();
+            //           break;               
+            // 
+            // MOTIVATION:
+            // Table switch will perform two branches (range check and the actual computed jump).
+            // Computed jump costs the same or more expensive than a regular conditional branch.
+            // So compbined cost would be 2+ conditional branches.
+            // Based on benchmarks it is actually closer to 3
+            //
+            // On the other hand, handling a degenerate bucket requires only 1 compare (range check) 
+            // so crumbling one bucket into two degenerate ones, when possible, is advantageous.
+            //
+            // One relatively common case that may result in degenerate switches is in async/iterator 
+            // state machines when control flow must use cascading dispatch into try blocks
+            internal int DegenerateBucketSplit
```

Comment:
The input is a "dense" bucket - the one that previous heuristics determined as not worth splitting, It has rough execution cost of a 1 regular branch (range check) and 1 computed branch (which cost roughly the same or perhaps more).
The only way to surely beat that cost via splitting is if the bucket can be split into 2 degenerate buckets. Then we have 2 regular branches.

3 degenerate buckets would require up to 3 regular compares. On some hardware computed jumps may cost significantly more than regular ones (because they are harder to predict or whatever), so it could still be profitable, but I did not want to guess that.

Basically if we have 3 degenerate buckets that can be merged into a dense bucket, we prefer a dense bucket, which we emit as "switch" opcode.

---
In reply to: [93716855](https://github.com/dotnet/roslyn/pull/16054#discussion_r93716855) [](ancestors = 93716855)